sharding/contracts/validator_manager.sol

```diff
@@ -0,0 +1,356 @@
+    emptySlotsStackTop = 0;
+    depositSize = 100 ether;
+    shufflingCycleLength = 5; // FIXME: just modified temporarily for test;
+    sigGasLimit = 400000;
+    periodLength = 5;
+    numValidatorsPerCycle = 100;
+    shardCount = 100;
+    addHeaderLogTopic = keccak256("add_header()");
+    sighasher = SigHasherContract(0xDFFD41E18F04Ad8810c83B14FD1426a82E625A7D);
+  }
+
+  function isStackEmpty() internal view returns(bool) {
+    return emptySlotsStackTop == 0;
+  }
+  function stackPush(int index) internal {
+    emptySlotsStack[emptySlotsStackTop] = index;
+    ++emptySlotsStackTop;
+  }
+  function stackPop() internal returns(int) {
+    if (isStackEmpty())
+      return -1;
+    --emptySlotsStackTop;
+    return emptySlotsStack[emptySlotsStackTop];
+  }
+
+  function getValidatorsMaxIndex() public view returns(int) {
+    address zeroAddr = 0x0;
+    int activateValidatorNum = 0;
+    int currentCycle = int(block.number) / shufflingCycleLength;
+    int allValidatorSlotsNum = numValidators + emptySlotsStackTop;
+
+    // TODO: any better way to iterate the mapping?
+    for (int i = 0; i < 1024; ++i) {
+        if (i >= allValidatorSlotsNum)
+            break;
+        if ((validators[i].validationCodeAddr != zeroAddr) &&
+            (validators[i].cycle <= currentCycle))
+            activateValidatorNum += 1;
+    }
+    return activateValidatorNum + emptySlotsStackTop;
+  }
+
+  function deposit(ValidatorContract _validationCodeAddr, address _returnAddr) public payable returns(int) {
+    require(!isValcodeDeposited[_validationCodeAddr]);
+    require(msg.value == depositSize);
+    // Find the empty slot index in validators set
+    int index;
+    int nextCycle = 0;
+    if (!isStackEmpty())
+      index = stackPop();
+    else {
+      index = int(numValidators);
+      nextCycle = (int(block.number) / shufflingCycleLength) + 1;
+      validators[index] = Validator({
+        deposit: msg.value,
+        validationCodeAddr: _validationCodeAddr,
+        returnAddr: _returnAddr,
+        cycle: nextCycle
+      });
+    }
+    ++numValidators;
+    isValcodeDeposited[_validationCodeAddr] = true;
+    
+    log2(keccak256("deposit()"), bytes32(address(_validationCodeAddr)), bytes32(index));
+    return index;
+  }
+
+  function withdraw(int _validatorIndex, bytes32 _sig) public returns(bool) {
+    var msgHash = keccak256("withdraw");
+    // TODO: Check this function
+    var result = validators[_validatorIndex].validationCodeAddr.withdraw.gas(sigGasLimit)(_sig);
+    if (result) {
+      validators[_validatorIndex].returnAddr.transfer(validators[_validatorIndex].deposit);
+      isValcodeDeposited[validators[_validatorIndex].validationCodeAddr] = false;
+      delete validators[_validatorIndex];
+      stackPush(_validatorIndex);
+      --numValidators;
+      log1(msgHash, bytes32(_validatorIndex));
+      return result;
+    }
+  }
+
+  function sample(int _shardId) public constant returns(address) {
```

Comment:
Is this meant to be getEligibleProposer?

Reply:
I think so, ref: https://github.com/ethereum/py-evm/blob/sharding/evm/vm/forks/sharding/contracts/validator_manager.v.py#L169